File: projects/python/dao/dao.py
#!/usr/bin/env python3
"""
DAO - Data Access Object providing unified interface 🐍

All operations follow an async pattern:
1. Generate UUID immediately
2. Store result with status="pending"
3. Execute operation (simulated synchronously for demo)
4. Update result to status="completed"
5. Return initial response {id: uuid, status: "pending"}

Clients poll for results using /result/poll operation.
Sssso elegant! 🐍
"""

import uuid
import sys
from pathlib import Path
from typing import Dict, Any

# Add parent directory to path to import store and operations
sys.path.insert(0, str(Path(__file__).parent.parent))

from store import ResultStore
from operations import prompt


class DAO:
    """
    Unified DAO interface for all operations 🐍

    Provides call(operation, params) method that routes to handlers.
    """

    def __init__(self):
        """Initialize DAO with result store 🐍"""
        self.store = ResultStore()

    def call(self, operation: str, params: Dict[str, Any] = None) -> Dict[str, Any]:
        """
        Call an operation with parameters 🐍

        Args:
            operation: Hierarchical operation path (e.g., "/prompt/generate")
            params: Operation-specific parameters

        Returns:
            Operation response with id, status, and optional result or error
        """
        if params is None:
            params = {}

        # Special case: /result/poll retrieves existing result 🐍
        if operation == '/result/poll':
            poll_id = params.get('id')
            if not poll_id:
                return {'error': 'Missing required parameter: id'}

            stored = self.store.get(poll_id)
            if not stored:
                return {'error': f'Result not found or expired: {poll_id}'}

            return stored

        # Generate correlation UUID 🐍
        id = str(uuid.uuid4())

        # Route to operation handler
        try:
            result = self._route_operation(operation, params)

            # Special case: /usage/track completes synchronously 🐍
            if operation == '/usage/track':
                completed_response = {
                    'id': id,
                    'status': 'completed',
                    'result': result
                }
                self.store.set(id, completed_response)
                return completed_response

            # Standard async pattern: store completed result, return pending 🐍
            completed_response = {
                'id': id,
                'status': 'completed',
                'result': result
            }
            self.store.set(id, completed_response)

            # Return initial pending response (async pattern) 🐍
            return {
                'id': id,
                'status': 'pending'
            }

        except Exception as e:
            # Store failed result and return error 🐍
            failed_response = {
                'id': id,
                'status': 'failed',
                'error': str(e)
            }
            self.store.set(id, failed_response)

            return {'error': str(e)}

    def _route_operation(self, operation: str, params: Dict[str, Any]) -> Dict[str, Any]:
        """
        Route operation to appropriate handler 🐍

        Args:
            operation: Operation path
            params: Operation parameters

        Returns:
            Operation result

        Raises:
            ValueError: If operation is unknown
        """
        if operation == '/prompt/generate':
            return prompt.process_prompt(params)
        elif operation == '/template/create':
            return prompt.create_template(params)
        elif operation == '/template/render':
            return prompt.render_template(params)
        elif operation == '/usage/track':
            return prompt.track_usage(params)
        else:
            raise ValueError(f'Invalid operation: {operation}')

File: projects/go/dao/dao.go
package dao

import (
	"fmt"

	"github.com/ahoward/tc/projects/go/operations"
	"github.com/ahoward/tc/projects/go/store"
)

// DAO provides a unified interface for operation invocation
type DAO struct {
	store *store.ResultStore
}

// NewDAO creates a new DAO instance
func NewDAO() *DAO {
	return &DAO{
		store: store.NewResultStore(),
	}
}

// Call invokes an operation with parameters
func (d *DAO) Call(operation string, params map[string]interface{}) map[string]interface{} {
	// Special case: /result/poll retrieves existing result
	if operation == "/result/poll" {
		pollID, ok := params["id"].(string)
		if !ok || pollID == "" {
			return map[string]interface{}{
				"error": "Missing required parameter: id",
			}
		}

		stored, exists := d.store.Get(pollID)
		if !exists {
			return map[string]interface{}{
				"error": fmt.Sprintf("Result not found or expired: %s", pollID),
			}
		}

		return d.responseToMap(stored)
	}

	// Generate correlation UUID
	id := operations.GenerateUUID()

	// Route to operation handler
	result, err := d.routeOperation(operation, params)

	if err != nil {
		// Store failed result
		failedResponse := store.OperationResponse{
			ID:     id,
			Status: "failed",
			Error:  err.Error(),
		}
		d.store.Set(id, failedResponse)

		return map[string]interface{}{
			"error": err.Error(),
		}
	}

	// Special case: /usage/track completes synchronously
	if operation == "/usage/track" {
		completedResponse := store.OperationResponse{
			ID:     id,
			Status: "completed",
			Result: result,
		}
		d.store.Set(id, completedResponse)

		return d.responseToMap(completedResponse)
	}

	// Standard async pattern: store completed result, return pending
	completedResponse := store.OperationResponse{
		ID:     id,
		Status: "completed",
		Result: result,
	}
	d.store.Set(id, completedResponse)

	// Return initial pending response
	return map[string]interface{}{
		"id":     id,
		"status": "pending",
	}
}

// routeOperation routes operation to appropriate handler
func (d *DAO) routeOperation(operation string, params map[string]interface{}) (map[string]interface{}, error) {
	switch operation {
	case "/prompt/generate":
		return operations.ProcessPrompt(params)
	case "/template/create":
		return operations.CreateTemplate(params)
	case "/template/render":
		return operations.RenderTemplate(params)
	case "/usage/track":
		return operations.TrackUsage(params)
	default:
		return nil, fmt.Errorf("Invalid operation: %s", operation)
	}
}

// responseToMap converts OperationResponse to map for JSON encoding
func (d *DAO) responseToMap(resp store.OperationResponse) map[string]interface{} {
	m := map[string]interface{}{
		"id":     resp.ID,
		"status": resp.Status,
	}

	if resp.Result != nil {
		m["result"] = resp.Result
	}

	if resp.Error != "" {
		m["error"] = resp.Error
	}

	return m
}
